### DISCUSSION TITLE
swiper/react problem in ssr mode

### DISCUSSION THREAD

**Discussion by matin-mortazavi** [nolimits4web/swiper#7340] [General]
### Check that this is really a bug

- [X] I confirm

### Reproduction link

hhtps://i-provide-code-in-description

### Bug description

hi guys i’ve a problem with swiper/react package in ssr mode

i got an parent component that fetch data and pass it to a component 
the child component show product in slider but problem is in first 0.5 second of first renedr all styles are messed up (in csr work well)
have you any idea how to handle it?! i think the problem related to slow loading of react-swiper

my parent component =>

 ```
export default async function Home() {
  const { data, categoryBox } = await coreUseCase.getBaseData();
  const {
      topCenterSlider,
     } = data;
  return (
    <div className={styles.home}>
      <TopSlider list={topCenterSlider} />
</div>  
```

my child component =>

```
import { SwiperSlide } from "swiper/react";
import { NextLink } from "@/components/Common";
import { Btn, Slider } from "@/components/Common";
import styles from "./TopSlider.module.scss";
import { Img } from "@/components/Common";
import { coreConstant } from "@/constant";
const sliderSettings = {
  navigation: false,
  loop: true,
  autoplay: {
    delay: 3000,
    disableOnInteraction: false,
  },
};

export default function TopSlider({ list }) {
  return (
    <div className={styles["top-slider"]}>
      <Slider settings={sliderSettings}>
        {list
          ?.filter((fil) => fil?.isEnable)
          ?.map((item, key) => (
            <SwiperSlide key={key}>
                           <figure className={styles["top-slider__img"]}>
                  <Img alt={item?.title} path={item?.image} category={coreConstant.Sliders} />
                </figure>
                <div className={styles["top-slider__details"]}>
                  <p className={styles["top-slider__details-text"]}>{item?.metaTitle}</p>
                </div>
                       </SwiperSlide>
          ))}
      </Slider>
    </div>
  );
}
```
my slider component =>
``` 
import { Swiper } from "swiper/react";
import { Navigation, Pagination, Scrollbar, A11y, Keyboard, Thumbs, Autoplay } from "swiper/modules";

import "swiper/css/bundle";
import "swiper/css";

const Slider = ({ settings, children }) => {
  const settingConfigs = {
    modules: [Navigation, Pagination, Scrollbar, A11y, Keyboard, Thumbs, Autoplay],
    pagination: { clickable: true },
    navigation: true,
    loop: true,
    ...settings,
  };

  return <Swiper {...settingConfigs}>{children}</Swiper>;
};

export default Slider;
```

### Swiper version

11.0.4

### Platform/Target and Browser Versions

chorome 

### Validations

- [X] Follow our [Code of Conduct](https://github.com/nolimits4web/swiper/blob/master/CODE_OF_CONDUCT.md)
- [X] Read the [docs](https://swiperjs.com/swiper-api).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.
- [X] Make sure this is a Swiper issue and not a framework-specific issue

### Would you like to open a PR for this bug?

- [ ] I'm willing to open a PR

**Comment by s-patompong** [comment]
We face the same issue, on initial render (for a split second), the style messed up. It'd be nice to have swiper working nice with SSR.

**Reply by mahmoudAcm** [reply]
The issue arises because Swiper calculates and assigns each slide's width and margin only after the JavaScript loads. This can cause layout shifts, especially in server-side rendering (SSR) environments where initial styles are applied inconsistently. 

To prevent this, a workaround could involve pre-setting the slide widths and margins in CSS based on the expected layout, so the layout is more stable before Swiper's JavaScript adjusts the slides. Additionally, if the layout shifts are too significant, consider implementing a custom solution to handle these styles on the server side.

**Reply by mahmoudAcm** [reply]
I had to implement a workaround using `<style>` tags to replicate the same width and margin calculations that Swiper applies to each slide. By pre-calculating and setting these styles, I could ensure layout stability even before Swiper’s JavaScript fully loads. This approach helped to avoid layout shifts, especially under server-side rendering (SSR) conditions, by mimicking Swiper’s dynamic styling process upfront.

**Comment by mateusp7** [comment]
Someone fix this issue?

**Reply by mahmoudAcm** [reply]
@mateusp7  Yes, I did! I'm actually really proud, even though the solution was surprisingly easy to implement.

**Reply by mahmoudAcm** [reply]
If you are interested just ping me

**Reply by sylvainJacot** [reply]
@mahmoudAcm Hey! I'm interested about your workaround using `<style>`. Or you maybe found a better solution. I'm wondering if I will not swap to another lib that supports SSR natively

**Reply by mahmoudAcm** [reply]
This line defines a unique ID for each instance of a `swiper` component, ensuring style isolation when there are nested or multiple `swiper` instances on the same page.

```javascript
const id = 'swiper-' + useId().replace(/:/g, '');
```

This line sets up the `_breakpoints` object to define how many slides are shown per view at different screen widths for the `swiper` component. 

```javascript
const _breakpoints = { 0: { slidesPerView: 1 }, ...breakpoints };
```

This code generates responsive CSS styles for each defined breakpoint by iterating over `_breakpoints` and dynamically creating a unique style rule for each screen width.

```javascript
const _className = Object.entries(_breakpoints)
  .map(([key, value]) => {
    const { slidesPerView = 1, spaceBetween: gap = spaceBetween } = value;
    return `
       @media (min-width: ${key}px) {
          .${id}-slide { 
            width: calc((100% - ${(slidesPerView - 1) * gap}px) / ${slidesPerView});
            margin-right: ${gap}px;
          }            
        }`;
  })
  .join(' ')
  .replace(/\s+/g, ' ');
```

This line inserts the dynamically generated CSS into the document by embedding `_className` inside a `<style>` tag.

```javascript
<style>{_className}</style>
```
Each swiper slide should have the class `${id}-slide`, and this class depends on the unique `id` generated in the first step. This ensures that all slides within the same swiper instance share the same class, allowing them to be styled consistently based on the dynamically generated `id`. This approach prevents style conflicts when multiple swiper instances are nested.

**Reply by mahmoudAcm** [reply]
@sylvainJacot if nothing is clear ping me!

**Reply by diazevedo** [reply]
Hey @mahmoudAcm , amazing work, thanks for sharing it.

so basically you first set in CSS what swiper will do after js loads. Is that right?

**Reply by mahmoudAcm** [reply]
@diazevedo Yes, you first define the CSS for the swiper's layout and appearance before JavaScript kicks in. Once JS loads, it adds interactive functionality on top of that CSS.

**Reply by diazevedo** [reply]
amazing, thank you very much! I'll try that one, at the moment another issue is to make Swiper respect the min-width of slides children.  Any idea? 
thank you thank you

**Reply by mahmoudAcm** [reply]
@diazevedo can you show me an example?

**Reply by diazevedo** [reply]
sure @mahmoudAcm.
basically, the tiles have a min-width of 336px but on a screen which should all be fine swiper sets the slide to 322px as in the example, it cuts the last slide and I think it is because the maths are wrong. Screenshots have min-width as max-content but it does not work too. 
I've tried different combinations, gap, margin, padding, CSS mode true...

**Reply by mahmoudAcm** [reply]
@diazevedo can you tell me the final result that you want to achieve?

**Reply by diazevedo** [reply]
@mahmoudAcm  thanks for taking the time to look into this. I much appreciate it!
So, I wanted the carousel to follow the paddings I have on the screen.
Let's say 20px on each side. But when my items have a min-width the calculation swiper makes the width of the slides go smaller than the min-width of tiles, this causes the last element to be cut off a bit.

something like this [codepen with example](https://codepen.io/redrobot753/pen/NWEowaL?editors=1100)

**Reply by mahmoudAcm** [reply]
@diazevedo  You gave the tail a min-width of 21rem and adding padding to each tile am i right?

**Reply by diazevedo** [reply]
@mahmoudAcm I'll try it today, thanks a lot!

**Reply by yunom-io** [reply]
@mahmoudAcm you're solution is simple and amazing, thank you!
